### ISSUE TITLE
Strings freed prematurely when used in a loop

### ISSUE THREAD

**Issue by starsiderfox** [nim-works/nimskull#1385]
### Specification

Strings shouldn't be freed prematurely when used in a loop.

### Example

    proc f(s: cstring) =
      var x = ""
      for i in 0..1:
        let y = $s
        echo x == y
        x = y
    let s = "x"
    f(s.cstring)

Run this code with `nim r -d:useMalloc --passC:-fno-omit-frame-pointer --passC:-mno-omit-leaf-frame-pointer --passC:-fsanitize=address --passL:-fsanitize=address test.nim`

### Actual Output

According to the address sanitizer, the string `y` is used after free (while allocating another one I suppose).

### Expected Output

    false
    true

### Possible Solution

zerbina says it's a bug with cursor inference, and my guess is that when it's being assigned to another variable that it should either disable the inference or to make a copy of the value. If the value is not going to be used again in the loop, maybe it could just swap the string value contents (pointer to payload, length).

**Comment by zerbina** [comment]
The bug is not directly related to `string` -- it affects all types that have custom
destruction/copy logic. A reproducer of the same bug that doesn't rely on ASAN:
```nim
type Object = object
  has: bool

var destroyed: int

proc `=destroy`(x: var Object) =
  if x.has:
    inc destroyed

proc use(x: Object) = discard

proc test() =
  var x: Object
  var i = 0
  while i == 0: # only run the body a single time
    let y = Object(has: true)
    use x
    x = y
    inc i
  
  # `y` was moved into `x`, so no initialized ``Object`` must have been
  # destroyed at this point
  doAssert destroyed == 0

test()
```

**Edit:** I've update the test case. In it's previous form, a cursor wouldn't have been harmful, since there was usage observing `x` borrowing from a dead location.

`x` is currently erroneously turned into a cursor, meaning that the `x = y` assignment
is a shallow assignment, with `y` being destroyed immediately afterwards.

**Comment by zerbina** [comment]
Fixed by #1388.
